### ISSUE TITLE
pytorch version not compatible with k2

### ISSUE THREAD

**Issue by tz301** [lhotse-speech/lhotse#399]
pip3 install lhotse will install torch=1.9.0，but k2 doesn't support this version. How to solve this?

**Comment by csukuangfj** [comment]
You can install PyTorch first. After that, you can install lhotse, which will reuse the locally installed PyTorch.

k2 supports torch 1.9.0 if you use conda to install it

**Comment by csukuangfj** [comment]
Or you can install k2 first and then install lhotse.

pip install k2 

pip install lhotse

**Comment by csukuangfj** [comment]
Please see 
https://icefall.readthedocs.io/en/latest/installation/index.html

**Comment by tz301** [comment]
> Please see
> https://icefall.readthedocs.io/en/latest/installation/index.html

Hi, that's exactly what I'm following. I try install k2 first, see below.
<img width="1416" alt="WX20210904-200801" src="https://user-images.githubusercontent.com/35760938/132093918-22c0a44e-08b4-44a1-9446-78beccba21a0.png">

**Comment by csukuangfj** [comment]
Please run
```bash
pip uninstall torchaudio
pip install lhotse
```

You have installed torchaudio==0.9.0, which depends on torch==1.9.0.
To reuse the torch installed by k2, you have to uninstall torchaudio first. Otherwise, torchaudio will upgrade
torch to 1.9.0

**Comment by tz301** [comment]
> Please run
> 
> ```shell
> pip uninstall torchaudio
> pip install lhotse
> ```
> 
> You have installed torchaudio==0.9.0, which depends on torch==1.9.0.
> To reuse the torch installed by k2, you have to uninstall torchaudio first. Otherwise, torchaudio will upgrade
> torch to 1.9.0

Yeah, I see, pip install lhotse will automatically install torchaudio==0.9.0. Even if I uninstall torchaudio and reinstall lhotse, it will install torchaudio==0.9.0 again. This is how I solve.

```
pip install lhotse
pip install k2
pip uninstall torchaudio
pip install torchaudio==0.7.2
```

Then yesno example in icefall project is ok to run.

I think this is still a issue, right? The pip also show some warning:

![image](https://user-images.githubusercontent.com/35760938/132094716-49fd5f20-39fb-4640-8b48-5a93a803eeb1.png)

**Comment by danpovey** [comment]
Looks like lhotse's requirements should depend on torch's version:
```
    torch_ver = LooseVersion(torch.__version__)
    if torch_ver >= LooseVersion("1.9.1"):
        raise NotImplementedError("Not yet supported")
    elif torch_ver >= LooseVersion("1.9.0"):
        install_requires.append("torchaudio==0.9.0")
    elif torch_ver >= LooseVersion("1.8.1"):
        install_requires.append("torchaudio==0.8.1")
    elif torch_ver >= LooseVersion("1.8.0"):
        install_requires.append("torchaudio==0.8.0")
    elif torch_ver >= LooseVersion("1.7.1"):
        install_requires.append("torchaudio==0.7.2")
    else:

```
from lhotse's setup.py.  This may depend on what torch was active when lhotse was installed originally.

**Comment by csukuangfj** [comment]
>  Even if I uninstall torchaudio and reinstall lhotse, it will install torchaudio==0.9.0 again. This is how I solve.

Did you uninstall lhotse and re-install it?

```
pip uninstall torchaudio
pip uninstall lhotse
pip install --no-cache-dir lhotse
```
should work, I believe.

**Comment by tz301** [comment]
> > Even if I uninstall torchaudio and reinstall lhotse, it will install torchaudio==0.9.0 again. This is how I solve.
> 
> Did you uninstall lhotse and re-install it?
> 
> ```
> pip uninstall torchaudio
> pip uninstall lhotse
> pip install --no-cache-dir lhotse
> ```
> 
> should work, I believe.

Yeah, seems work.

**Comment by tz301** [comment]
I reopen this since I still meet this issue when I install lhotse using: `pip install lhotse`. torchaudio==0.9.0 will be installed which is not compatible with my torch version (1.8.1), and then reinstall torch=1.9.0 automatically.

But if I clone this project and run `python3 setup.py install`, torchaudio==0.8.1 will be installed which is compatible with my torch version == 1.8.1. The code in `setup.py` seems working this way:
```
    torch_ver = LooseVersion(torch.__version__)
    if torch_ver >= LooseVersion("1.9.1"):
        raise NotImplementedError("Not yet supported")
    elif torch_ver >= LooseVersion("1.9.0"):
        install_requires.append("torchaudio==0.9.0")
    elif torch_ver >= LooseVersion("1.8.1"):
        install_requires.append("torchaudio==0.8.1")
    elif torch_ver >= LooseVersion("1.8.0"):
        install_requires.append("torchaudio==0.8.0")
    elif torch_ver >= LooseVersion("1.7.1"):
        install_requires.append("torchaudio==0.7.2")
    else:
```

It seems the version check in pip release package not working.

**Comment by csukuangfj** [comment]
By the way, what is the order that you used to install k2, lhotse, torch, and torchaudio?

**Comment by csukuangfj** [comment]
One thing you need to check is that once you installed k2, you cannot change your PyTorch version.
If you change that, you have to re-install k2 to match your local PyTorch version, since k2 is compiled against
a specific version of PyTorch

**Comment by tz301** [comment]
> By the way, what is the order that you used to install k2, lhotse, torch, and torchaudio?

1. I have torch=1.8.1 in my environment.
2. I install k2 by clone it and run `python3 setup.py install`, so that no compatibility problem between k2 and torch.
3. Install lhotse (no torchaudio currently). I try two diffenent ways to install:
(1) `pip install lhotse`: this will unexpectedly try to install `torchaudio==0.9.0` and will try to reinstall torch.
(2) clone lhotse and run `python3 setup.py install`, this will install `torchaudio==0.8.1` correctly, and torch will not be reinstalled.

So I doubt torchaudio version check not working in pip release package.

**Comment by csukuangfj** [comment]
> So I doubt torchaudio version check not working in pip release package.

Maybe @pzelasko has more to say about this.

**Comment by pzelasko** [comment]
Hmm, managing the right torchaudio version definitely seems tricky..

I don't understand why it selected `torchaudio==0.9.0` out of all versions; if it went for the latest, it'd install 0.10.0 instead. This is the relevant `setup.py` code in latest pip released version (0.11); I can't see what's going wrong. 

https://github.com/lhotse-speech/lhotse/blob/6d60f333da89abc3c2836f395f95ae9d53ca5cd0/setup.py#L150-L171

Maybe I can add some logging utils for this controlled via env var (`LHOTSE_DEBUG_INSTALL=1 pip install lhotse`) so that we can debug it more easily after the next release..

**Comment by pzelasko** [comment]
This is solved by #465 -- as of now, Lhotse will require you to have either both torch and torchaudio installed, or none of them installed. The burden of ensuring the right torchaudio version is on the user (but we provide an error message links where you can find the info how to install it).
